kubetest/anywhere.go

```diff
@@ -199,3 +205,126 @@ func (k kubernetesAnywhere) Down() error {
 }
+
+const DefaultConfigFile = ".config"
+
+type kubernetesAnywhereMulti struct {
+	*kubernetesAnywhere
+	multiClusters multiClusterDeployment
+}
+
+var _ deployer = kubernetesAnywhereMulti{}
+
+func newKubernetesAnywhereMulti(project, zone string, multiClusters multiClusterDeployment) (*kubernetesAnywhereMulti, error) {
+	if len(multiClusters.clusters) < 1 {
+		return nil, fmt.Errorf("invalid --multi-clusters flag passed")
+	}
+	k, _ := newKubernetesAnywhere(project, zone)
+	mk := &kubernetesAnywhereMulti{k, multiClusters}
+
+	for _, cluster := range mk.multiClusters.clusters {
+		specificZone, specified := mk.multiClusters.zones[cluster]
+		if specified {
+			mk.Zone = specificZone
+		} else {
+			mk.Zone = zone
+		}
+		mk.Cluster = cluster
+		mk.KubeContext = mk.Zone + "-" + mk.Cluster
+		if err := mk.writeConfig(); err != nil {
+			return nil, err
+		}
+	}
+	return mk, nil
+}
+
+func (k kubernetesAnywhereMulti) writeConfig() error {
+	config, err := k.getConfig()
+	if err != nil {
+		return fmt.Errorf("Could not generate config: %v", err)
+	}
+
+	configFileName := DefaultConfigFile + "-" + k.Cluster
+	f, err := os.Create(k.path + "/" + configFileName)
+	if err != nil {
+		return fmt.Errorf("Could not create file: %v", err)
+	}
+	defer f.Close()
+
+	fmt.Fprint(f, config)
+	return nil
+}
+
+func (k kubernetesAnywhereMulti) Up() error {
+	var cmds []*exec.Cmd
+	for _, cluster := range k.multiClusters.clusters {
+		configFile := DefaultConfigFile + "-" + cluster
+		cmd := exec.Command("make", "-C", k.path, "CONFIG_FILE="+configFile, "deploy")
+		cmds = append(cmds, cmd)
+	}
+
+	if err := finishRunningParallel(cmds...); err != nil {
+		return err
+	}
+
+	return k.TestSetup()
+}
+
+func (k kubernetesAnywhereMulti) TestSetup() error {
+	var kubecfg string
+	for _, clusterName := range k.multiClusters.clusters {
+		configFile := DefaultConfigFile + "-" + clusterName
+		o, err := output(exec.Command("make", "--silent", "-C", k.path, "CONFIG_FILE="+configFile, "kubeconfig-path"))
+		if err != nil {
+			return fmt.Errorf("Could not get kubeconfig-path: %v", err)
+		}
+		if len(kubecfg) != 0 {
+			kubecfg += ":"
+		}
+		kubecfg += strings.TrimSuffix(string(o), "\n")
+	}
+
+	if err := os.Setenv("KUBECONFIG", kubecfg); err != nil {
+		return err
+	}
+	return nil
+}
+
+func (k kubernetesAnywhereMulti) IsUp() error {
+	if err := k.TestSetup(); err != nil {
+		return err
+	}
+
+	for _, clusterName := range k.multiClusters.clusters {
+		zone := k.multiClusters.zones[clusterName]
+		kubeContext := zone + "-" + clusterName
+		o, err := output(exec.Command("kubectl", "--context="+kubeContext, "get", "nodes", "--no-headers"))
+		if err != nil {
+			log.Printf("kubectl get nodes failed for cluster %s: %s\n%s", clusterName, wrapError(err).Error(), string(o))
+			return err
+		}
+		stdout := strings.TrimSpace(string(o))
+		log.Printf("Cluster nodes of cluster %s:\n%s", clusterName, stdout)
+
+		n := len(strings.Split(stdout, "\n"))
+		if n <= 0 {
```

Comment:
Does it matter to wait until `n` becomes the number of expected nodes? It only looks like you wait until any nodes are found.

Reply:
initially i thought, if the cluster is up with just one node it should be fine to continue with further operations. It doesn't harm to wait for all nodes to be ready. So fixed it according to the suggestion.